### ISSUE TITLE
Deadlock on dispose

### ISSUE THREAD

**Issue by AnthonyMillerSF** [DeclarativeHub/ReactiveKit#236]
I just updated from 3.13.0 to 3.14.2. In my unit tests I'm getting sporadic dead locks on disposing a bag in the `tearDown()` method of my tests. 

I have a pretty complicated setup, using `retry` and a bunch of composite disposables. But I've got an extensive unit test suite, in `3.13.0`, all my unit tests passed consistently.

Is it possible that I have an error in my unit tests? Even so, I don't think it should be causing a dead lock now. I'm wondering if there is a bug in the new threading/locking code in `3.14.2`.

**Comment by srdanrasic** [comment]
Hmm, it’s possible. Do you have any more info that could help us trace the problem?

**Comment by AnthonyMillerSF** [comment]
I've been digging into it all day. This is my current understanding, may not be fully accurate:

It looks like this may be deadlocking when one thread is trying to observe a `Signal` while another thread is disposing of it. 

The `AtomicObserver.lock` is the lock that is deadlocking I think. 

This may have something to do with an interaction between
`public init<S: Sequence>(sequence: S, interval: Double, queue: DispatchQueue = DispatchQueue(label: "com.reactive_kit.signal.sequence"))`
and 
`public func retry<S: SignalProtocol>(when other: S, if shouldRetry: @escaping (Error) -> Bool`


Trying to verify and get more information. Will update more as I find info.

**Comment by AnthonyMillerSF** [comment]
Alright, I've added the following print statements to `AtomicObserver`:
```
/// Observer that ensures events are sent atomically.
public final class AtomicObserver<Element, Error: Swift.Error>: ObserverProtocol {
    
    private var observer: Observer<Element, Error>?
    private let lock = NSRecursiveLock(name: "com.reactive_kit.atomic_observer")
    private let parentDisposable: Disposable
    
    private var _disposable: Disposable?
    public var disposable: Disposable {
        print("\(Unmanaged.passUnretained(self).toOpaque()) Access disposable LOCK - \(Thread.current)")
        lock.lock();
        defer {
            lock.unlock()
            print("\(Unmanaged.passUnretained(self).toOpaque()) Access disposable UNLOCK - \(Thread.current)")
        }
        return _disposable!
    }
    
    /// Creates an observer that wraps given closure.
    public init(disposable: Disposable, observer: @escaping Observer<Element, Error>) {
        self.observer = observer
        self.parentDisposable = disposable

        print("\(Unmanaged.passUnretained(self).toOpaque()) Init LOCK - \(Thread.current)")
        lock.lock();
        defer {
            lock.unlock()
            print("\(Unmanaged.passUnretained(self).toOpaque()) Init UNLOCK - \(Thread.current)")
        }
        _disposable = BlockDisposable { [weak self] in
            if let strongSelf = self {
                print("\(Unmanaged.passUnretained(strongSelf).toOpaque()) Disposal Block LOCK - \(Thread.current)")
                strongSelf.lock.lock()
                strongSelf.observer = nil

                strongSelf.lock.unlock()
                print("\(Unmanaged.passUnretained(strongSelf).toOpaque()) Disposal Block UNLOCK - \(Thread.current)")
            }
            disposable.dispose()
        }
    }
    
    /// Calles wrapped closure with the given element.
    public func on(_ event: Signal<Element, Error>.Event) {
        print("\(Unmanaged.passUnretained(self).toOpaque()) on: LOCK - \(Thread.current)")
        lock.lock();
        defer {
            lock.unlock()
            print("\(Unmanaged.passUnretained(self).toOpaque()) on: UNLOCK - \(Thread.current)")
        }

        guard let disposable = _disposable, !disposable.isDisposed else { return }
        if let observer = observer {
            print("\(Unmanaged.passUnretained(self).toOpaque()) on: Start Event - \(Thread.current)")
            observer(event)
            print("\(Unmanaged.passUnretained(self).toOpaque()) on: End Event - \(Thread.current)")
            if event.isTerminal {
                print("\(Unmanaged.passUnretained(self).toOpaque()) on: Start Disposal - \(Thread.current)")
                disposable.dispose()
                print("\(Unmanaged.passUnretained(self).toOpaque()) on: End Disposal - \(Thread.current)")
            }
        }
    }
}
````


When I run and catch the dead lock, I get this log output:
```
0x00007fddb1b17940 Init LOCK - <NSThread: 0x7fddb1801470>{number = 1, name = main}
0x00007fddb1b17940 Init UNLOCK - <NSThread: 0x7fddb1801470>{number = 1, name = main}
0x00007fddb1b17940 Access disposable LOCK - <NSThread: 0x7fddb1801470>{number = 1, name = main}
0x00007fddb1b17940 Access disposable UNLOCK - <NSThread: 0x7fddb1801470>{number = 1, name = main}
0x00007fddb1b17940 on: LOCK - <NSThread: 0x7fddb0578ec0>{number = 5, name = (null)}
0x00007fddb1b17940 on: Start Event - <NSThread: 0x7fddb0578ec0>{number = 5, name = (null)}
0x00007fddb1b17940 on: End Event - <NSThread: 0x7fddb0578ec0>{number = 5, name = (null)}
0x00007fddb1b17940 on: UNLOCK - <NSThread: 0x7fddb0578ec0>{number = 5, name = (null)}
0x00007fddb1b17940 on: LOCK - <NSThread: 0x7fddb0578ec0>{number = 5, name = (null)}
0x00007fddb1b17940 on: Start Event - <NSThread: 0x7fddb0578ec0>{number = 5, name = (null)}
0x00007fddb1b17940 Disposal Block LOCK - <NSThread: 0x7fddb1801470>{number = 1, name = main}
```

The lock is being acquired by the `on:` function, which is then calling `observer(event)`. But that call never finishes and then it deadlocks.

This tells me that I probably have something going on where my observer triggers a chain of events which then disposes of the observer on the main thread? Unless I am dispatching `sync` to the main thread somewhere, I don't see why this should be a deadlock yet.

My thought is that there is likely, even if I am dispatching `sync` to the main thread, having this cause a deadlock is not the best behavior for the library to have.

Will update with more information if I find it.

**Comment by AnthonyMillerSF** [comment]
> This may have something to do with an interaction between
> `public init<S: Sequence>(sequence: S, interval: Double, queue: DispatchQueue = DispatchQueue(label: "com.reactive_kit.signal.sequence"))`
> and
> `public func retry<S: SignalProtocol>(when other: S, if shouldRetry: @escaping (Error) -> Bool`

This seems to be the case. I think I may have an idea of what could be happening. 

I have a function to retry a signal at a polling interval if it fails.

```
public struct PollInfo {
    var interval: Double
    var repeatCount: Double
}

extension SignalProtocol {

    func retry(for pollInfo: PollInfo) -> Signal<Element, Error> {
        let repeatCount = Int(pollInfo.repeatCount)
        guard repeatCount > 1 else { return self.toSignal() }

        return Signal { observer in
            let subject = Signal<Int, Never>(
                sequence: 1...repeatCount - 1,
                interval: pollInfo.interval
            ).publish()
            let retryDisposable = self.retry(when: subject).observe(with: observer.on)
            let pollingDisposable = subject.connect()

            return CompositeDisposable([retryDisposable, pollingDisposable])
        }
    }
}
```

I am turning the polling interval signal into a `Subject` and starting immediately. When the `self` fails, it is beginning an observer on the `subject` at `SignalProtocol+ErrorHandling.swift line 160`. When the polling interval triggers, it calls `Subject.on(_ event:)`. This acquires the lock. It then calls its observer, which retries the original signal (`self`). If the original signal fails again with the wrong timing, the `retry(when other: if:)` will be trying to observe and get its `AtomicObserver`'s disposable for `self.observe` and `other.observe` at the same time. This causes a deadlock in the `Subject`.

I'm 90% sure this is the cause. I don't know what the solution would be yet.

**Comment by AnthonyMillerSF** [comment]
When I comment out the `lock` in `Subject`, I no longer get this issue. 
Also, if I change `Subject` so that it unlocks prior to calling the `observers` in `on(_ event:)` it works as well.

I'm not sure what the value of locking around calling each of the `observers` is. We should be locking around MUTATING the `observers` for sure, but if we just make it an array of  `AtomicObserver`s, they are already each atomic themselves, so that should be enough.

Here is my proposal

Change `Subject.on(_ event:)` to this:
```swift
 public func on(_ event: Signal<Element, Error>.Event) {
        lock.lock();
        guard !isTerminated else { return }

        isTerminated = event.isTerminal

        let deletedObservers = self.deletedObservers.value
        observers.removeAll(where: { (token, _) in
            deletedObservers.contains(token)
        })
        self.deletedObservers.mutate {
            $0.subtracting(deletedObservers)
        }

        let observers = self.observers
        lock.unlock()

        for (_, observer) in observers {
            observer(event)
        }
    }
```

**Comment by AnthonyMillerSF** [comment]
Hmmm actually, it seems like that just moves the deadlock into the `retry` function's lock. Removing the lock on `Subject` altogether is the only thing that's worked so far...

**Comment by srdanrasic** [comment]
@AnthonyMDev thanks for trying to figure this out. I'd love to dig deeper into this, but I'm having trouble reproducing it. Do you think you could make a test case that can reproduce the issue?

**Comment by AnthonyMDev** [comment]
@srdanrasic 
I was able to create a simple test case the reproduces this every time! This is similar to what I'm doing in my production code. 
https://github.com/AnthonyMDev/ReactiveKitDeadlockDemo

If I'm doing something wrong in my code, let me know. If this is a bug in my code, we should probably still find a way to make this known in ReactiveKit by throwing an error or something, since we don't ever want people experiencing unexpected deadlocks.

**Comment by srdanrasic** [comment]
@AnthonyMDev, thank you, that test case is really helpful! I think I have an idea of what's causing the problem and I'll try to fix it tomorrow morning.

**Comment by srdanrasic** [comment]
Hi @AnthonyMillerSF,

I think I have resolved the threading issues. Would you be able to try the out the branch `feature/threading-improvements` in your project and see if that helps?

Btw, there is a small bug in your test app. A dispose bag is disposed before all signals get the chance to send their events so most of them will get stuck in `waitUntil` block even when there is no deadlock.

**Comment by AnthonyMillerSF** [comment]
Thanks for looking into this @srdanrasic. I fixed the bug in the demo test project I shared with you and then pulled in that branch. My test is still deadlocking with the same issue as before. :/

**Comment by srdanrasic** [comment]
Hmm, can you push the change you made @AnthonyMillerSF?

**Comment by AnthonyMillerSF** [comment]
Pushed up!

**Comment by srdanrasic** [comment]
@AnthonyMillerSF not sure if you've seen, but I left a comment on your commit: https://github.com/AnthonyMDev/ReactiveKitDeadlockDemo/commit/bb9bcf39972e03bb8cbc99329b74be74fe541104

**Comment by AnthonyMillerSF** [comment]
Just responded. I've made another change to the disposeBag that should alleviate your concerns. This dead lock is still showing up.

**Comment by srdanrasic** [comment]
@AnthonyMillerSF I can see that the unit test is not finishing, but if I pause the process I don't see any thread in a dead lock.

I've then looked into the retry implementation and I think that using the subject there might produce different behaviour that you want. If I change it to the following then I see no issues.

```swift
extension SignalProtocol {

    func retry(for pollInfo: PollInfo) -> Signal<Element, Error> {
        let repeatCount = Int(pollInfo.repeatCount)
        guard repeatCount > 1 else { return self.toSignal() }

        return Signal { observer in
            return self.retry(when: Signal<Int, Never>(
                sequence: 1...repeatCount - 1,
                interval: pollInfo.interval
            )).observe(with: observer.on)
        }
    }
}
```

**Comment by AnthonyMillerSF** [comment]
Hmmm. I'm looking at this more right now. I added some log statements, and it looks like it's not actually deadlocking in the test, but it is taking a long time to finish, and some of the test case iterations are occasionally taking longer than 1.0 seconds to finish, so they fail. Each iteration shouldn't take more than 0.4-0.5 seconds, so I'm not sure what is causing that. 

Let me try what you are reccomending here.

**Comment by AnthonyMillerSF** [comment]
When I change the `Signal` as you recommend, it does seem to work. My concern is that there still is possibly an issue in the locking on `Subjects`. I'm going to see if I can make a test case to prove it. 

If not, I'll close this issue as resolved. Thanks for your help so far. You and your repos are awesome @srdanrasic!

**Comment by srdanrasic** [comment]
Thank you @AnthonyMillerSF! If you find something, I'll be happy to look into it.

**Comment by AnthonyMillerSF** [comment]
So with the changes in `13.4.3`, my production code is now passing. 

> ...it looks like it's not actually deadlocking in the test, but it is taking a long time to finish.

I realized that wasn't true. The test iterations that are failing are deadlocking. I'm seeing their threads stuck there far after they should be gone, and I increased the timeout to 2 seconds. Still deadlocking with no sign of coming out of it.

The solution you gave me above (not using `publish()`) does work, but that just signifies to me that there is still the possibility of creating a deadlock in `Subject`. I've pushed up new changes in `master` to my demo test case for you. I hope it helps, but I'm still definitely getting a deadlock when using `Subjects`.
